### ISSUE TITLE
Performance toolkit exits before running any scenarios

### ISSUE THREAD

**Issue by sshymko** [magento/magento2#21106]
The [Magento Performance Toolkit](../tree/2.2.7/setup/performance-toolkit) is unable to run against a vanilla Magento 2 instance installed from scratch with the required fixture data. It successfully completes the preliminary preparation steps, but never gets to execute any frontend scenarios. JMeter exits without any errors.

### Preconditions (*)
1. Magento 2.2.7 w/ Redis session storage
```shell
mysql -e "CREATE DATABASE magento2"

git clone https://github.com/magento/magento2.git
cd magento2/
git checkout -b v2.2.7 2.2.7

composer install --no-dev
chmod -R ug+rwX app/etc/ var/ pub/media/
php bin/magento setup:install \
    --base-url="http://127.0.0.1/" \
    --backend-frontname="admin" \
    --use-rewrites=1 \
    --db-host="localhost" --db-name="magento2" \
    --admin-user="admin" --admin-password="123123q" \
    --admin-firstname="Admin" --admin-lastname="Admin" --admin-email="admin@example.com" \
    --language=en_US --currency=USD --timezone=America/Los_Angeles \
    --session-save=redis --session-save-redis-host=/var/run/redis.sock --session-save-redis-port=0 \
    --session-save-redis-db=0 --session-save-redis-disable-locking=1 --session-save-redis-log-level=3

php bin/magento config:set dev/static/sign 1
php bin/magento config:set dev/css/minify_files 1
php bin/magento config:set dev/js/minify_files 1
php bin/magento config:set dev/js/enable_js_bundling 1
php bin/magento config:set admin/security/admin_account_sharing 1
php bin/magento config:set admin/security/use_form_key 0

php bin/magento deploy:mode:set production
composer dump-autoload --no-dev --optimize

php bin/magento indexer:set-mode schedule
php bin/magento indexer:reindex

php bin/magento cache:flush
```
2. JMeter 3.1 w/ JSON plugin
```shell
sudo yum install java -y

curl -L -O https://archive.apache.org/dist/jmeter/binaries/apache-jmeter-3.1.tgz
tar -xzf apache-jmeter-3.1.tgz

curl -L -O https://jmeter-plugins.org/files/packages/jpgc-json-2.5.zip
unzip jpgc-json-2.5.zip -d ./apache-jmeter-3.1/
```

### Steps to reproduce (*)
1. Generate fixture data
```shell
php bin/magento setup:performance:generate-fixtures setup/performance-toolkit/profiles/ce/small.xml
```
2. Run JMeter test plan
```shell
jmeter -n \
    -t magento2/setup/performance-toolkit/benchmark.jmx \
    -j jmeter.log \
    -l jmeter-results.jtl \
    -Jhost=127.0.0.1 \
    -Jbase_path=/ \
    -Jadmin_path=admin \
    -Jadmin_user=admin \
    -Jadmin_password=123123q \
    -Jfiles_folder=magento2/setup/performance-toolkit/files/ \
    -Jloops=1 \
    -JfrontendPoolUsers=1 \
    -JadminPoolUsers=0 \
    -JcsrPoolUsers=0 \
    -JothersPoolUsers=0
```

### Expected result (*)
1. JMeter successfully runs all frontend scenarios

### Actual result (*)
1. JMeter successfully completes the preparation (log in to admin, load products, categories, customers, etc.)
2. JMeter exits before executing any frontend scenarios
```
Writing log file to: /home/ec2-user/jmeter.log
Creating summariser <summary>
Created the tree successfully using /home/ec2-user/magento2/setup/performance-toolkit/benchmark.jmx
Starting the test @ Sun Feb 10 05:51:07 UTC 2019 (1549777867357)
Waiting for possible Shutdown/StopTestNow/Heapdump message on port 4445
summary =    224 in 00:00:05 =   45.0/s Avg:    15 Min:     0 Max:   676 Err:     0 (0.00%)
Tidying up ...    @ Sun Feb 10 05:51:12 UTC 2019 (1549777872774)
... end of run
```
3. Nginx access logs shows only preparation requests
```
3.81.150.213 - - [10/Feb/2019:07:33:39 +0000] "GET /admin HTTP/1.1" 302 5 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:39 +0000] "GET /admin/admin/ HTTP/1.1" 200 2358 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:39 +0000] "POST /admin/admin/dashboard/ HTTP/1.1" 302 5 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:40 +0000] "GET /admin/admin/dashboard/ HTTP/1.1" 200 9246 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:40 +0000] "POST /rest/V1/integration/admin/token HTTP/1.1" 200 65 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:40 +0000] "GET /rest/V1/categories/list?searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bfield%5D=path&searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bvalue%5D=1%2F2%2F%25&searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5BconditionType%5D=like&searchCriteria%5BfilterGroups%5D%5B1%5D%5Bfilters%5D%5B0%5D%5Bfield%5D=level&searchCriteria%5BfilterGroups%5D%5B1%5D%5Bfilters%5D%5B0%5D%5Bvalue%5D=2&searchCriteria%5BpageSize%5D=100 HTTP/1.1" 200 536 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:40 +0000] "POST /rest/V1/integration/admin/token HTTP/1.1" 200 65 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:40 +0000] "GET /rest/default/V1/categories/list?searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bfield%5D=children_count&searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bvalue%5D=0&searchCriteria%5BfilterGroups%5D%5B1%5D%5Bfilters%5D%5B0%5D%5Bfield%5D=level&searchCriteria%5BfilterGroups%5D%5B1%5D%5Bfilters%5D%5B0%5D%5Bvalue%5D=2&searchCriteria%5BfilterGroups%5D%5B1%5D%5Bfilters%5D%5B0%5D%5BconditionType%5D=gt&searchCriteria%5BpageSize%5D=0 HTTP/1.1" 200 892 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:40 +0000] "POST /rest/V1/integration/admin/token HTTP/1.1" 200 65 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:40 +0000] "GET /rest/V1/products?searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bfield%5D=type_id&searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bvalue%5D=configurable&searchCriteria%5BpageSize%5D=15 HTTP/1.1" 200 8107 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:41 +0000] "POST /rest/V1/integration/admin/token HTTP/1.1" 200 65 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:41 +0000] "GET /rest/V1/products?searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bfield%5D=type_id&searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bvalue%5D=configurable&searchCriteria%5BpageSize%5D=15&searchCriteria%5BcurrentPage%5D=2 HTTP/1.1" 200 1271 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:41 +0000] "POST /rest/V1/integration/admin/token HTTP/1.1" 200 65 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:41 +0000] "GET /rest/V1/products?searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bfield%5D=type_id&searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bvalue%5D=simple&searchCriteria%5BpageSize%5D=30&searchCriteria%5BcurrentPage%5D=1&searchCriteria%5BfilterGroups%5D%5B1%5D%5Bfilters%5D%5B1%5D%5Bfield%5D=attribute_set_id&searchCriteria%5BfilterGroups%5D%5B1%5D%5Bfilters%5D%5B1%5D%5Bvalue%5D=4 HTTP/1.1" 200 5550 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:41 +0000] "POST /rest/V1/integration/admin/token HTTP/1.1" 200 65 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:41 +0000] "GET /rest/V1/products?searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bfield%5D=type_id&searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bvalue%5D=simple&searchCriteria%5BpageSize%5D=30&searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B1%5D%5Bfield%5D!=attribute_set_id&searchCriteria%5BfilterGroups%5D%5B0%5D%5Bfilters%5D%5B1%5D%5Bvalue%5D=4 HTTP/1.1" 200 5528 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:41 +0000] "POST /admin/admin/user/roleGrid/limit/200/?ajax=true&isAjax=true HTTP/1.1" 200 3209 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:42 +0000] "GET /admin/customer/index/ HTTP/1.1" 200 7914 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:42 +0000] "GET /admin/mui/index/render/?namespace=customer_listing&sorting%5Bfield%5D=entity_id&sorting%5Bdirection%5D=asc&isAjax=true&filters%5Bplaceholder%5D=customer_since%5Blocale%5D%3Den_US&filters%5Bgroup_id%5D=1&filters%5Bwebsite_id%5D=1&paging%5BpageSize%5D=100&paging%5Bcurrent%5D=1&sorting%5Bfield%5D=entity_id&sorting%5Bdirection%5D=asc&isAjax=true HTTP/1.1" 200 1932 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:42 +0000] "GET /admin/directory/json/countryRegion/?parent=US HTTP/1.1" 200 969 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:43 +0000] "POST /checkout/cart/add HTTP/1.1" 302 5 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:43 +0000] "GET / HTTP/1.1" 200 7050 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
3.81.150.213 - - [10/Feb/2019:07:33:43 +0000] "GET /page_cache/block/render/?blocks=[\x22customer_form_login\x22]&handles=[\x22default\x22,\x22customer_account_login\x22]&originalRequest={\x22route\x22:\x22customer\x22,\x22controller\x22:\x22account\x22,\x22action\x22:\x22login\x22,\x22uri\x22:\x22/customer/account/login/\x22}&ajax=true&isAjax=true HTTP/1.1" 200 754 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0" "-"
```
4. JMeter log shows no errors
[jmeter.log](https://github.com/magento/magento2/files/2848627/jmeter.log)
[jmeter-results.jtl](https://github.com/magento/magento2/files/2848628/jmeter-results.jtl.txt)

**Comment by magento-engcom-team** [comment]
Hi @sshymko. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@sshymko do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by sshymko** [comment]
@magento-engcom-team yes

**Comment by sshymko** [comment]
@kandy 
This is the issue with the performance toolkit I've been facing.
Any idea what could be wrong?

**Comment by pmouawad** [comment]
@sshymko can you attach the jmx file you use ? thanks

**Comment by sshymko** [comment]
@pmouawad 
This is an unmodified JMX file that comes with the performance toolkit:
https://github.com/magento/magento2/blob/2.2.7/setup/performance-toolkit/benchmark.jmx

**Comment by sshymko** [comment]
@pmouawad 
I see you're part of the JMeter team :)
Please note that the old version of JMeter 3.1 is used for testing.
More up-to-date versions 4 and 5 are unfortunately not supported.
I would really appreciate any clues of what could be wrong.

**Comment by pmouawad** [comment]
Thanks, any way to provide to me an access to your instance or a docker image or any other alternative ? thanks

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by sshymko** [comment]
@pmouawad 
Here's a test instance for you: http://3.85.65.228/
Feel free to play with it for the next ~10 hours; it will be terminated after that.
The JMeter command provided in the ticket description has all credentials embedded.
Thank you so much for looking into this issue, it's greatly appreciated.

**Comment by pmouawad** [comment]
It would be better if you let it live until 23h30 Paris Timezone, it’s 9am now. I am working for a customer, not sure I can dedicate time today before 20h30

**Comment by kandy** [comment]
@sshymko Looks like there is an issue of our Jmeter configuration. If loops count < 2 it will not run any scenario.
So current workaround will be to increase number of loops.

**Comment by sshymko** [comment]
@kandy 
Thank you so much! I'll try running with loops > 1 and let you know how it goes.

**Comment by sshymko** [comment]
@kandy 
I can confirm that the workaround of `-Jloops=2` does work!

**Comment by sshymko** [comment]
@kandy 
I'm just curious, how did you figure out that the issue was `loops=1`?
The usage of the `loops` parameter in the JMX is pretty straightforward and looks correct to me.
Is there a `loops=loops-1` logic hiding somewhere in the BinShell scripts?

**Comment by kandy** [comment]
@sshymko Looks like it's strange behavior in case **Thread Group** with **Throughput Controler** and throughput less than 100% work. The scenario with 100 throughput work with 1 loop too.

**Comment by sshymko** [comment]
@kandy 
Perhaps it's finally the time to upgrade to the latest JMeter that hopefully has no issues like that ;)

**Comment by pmouawad** [comment]
Hello,
The problem is that with a loop of 1, and none of the ThroughputController having 100%, you can end up executing nothing.
In the long run (more loops) the percentage will be matched.

This is expected behavior.
If you want a guarantee of running use one of those 2 approaches:
- https://github.com/Blazemeter/jmeter-bzm-plugins/blob/master/wsc/WeightedSwitchController.md
- https://www.ubik-ingenierie.com/blog/jmeter_control_percentage_of_sampler/

@kandy , if I can help on the test plan let me know where to make a kind of review.

I had reported following issue but AFAICU ,nothing was taken into account:

https://github.com/magento/magento2/issues/723

**Comment by pmouawad** [comment]
I see a potential issue in:
"SetUp - BeanShell Sampler: Collect simple product for edit"

on second iteration, editProductList is null
`
if (1 == Integer.parseInt(vars.get("simple_products_counter_for_edit"))) {
    editProductList = new ArrayList();
    props.put("simple_products_list_for_edit", editProductList);
} else {
    productList = props.get("simple_products_counter_for_edit");
}
`

Should it be ?

`
if (1 == Integer.parseInt(vars.get("simple_products_counter_for_edit"))) {
    editProductList = new ArrayList();
    props.put("simple_products_list_for_edit", editProductList);
} else {
    editProductList = props.get("simple_products_list_for_edit");
}

`

**Comment by pmouawad** [comment]
Also be aware that with upgrade of Beanshell version, it became more strict and you need to declare variables.

Anyway I suggest you move out of Beanshell, javascript which break scalability of JMeter

**Comment by sshymko** [comment]
All right, I'm shutting down the test instance 3.85.65.228.
I'll proceed with the workaround to unblock my testing.
Thanks everybody for the assistance!

**Comment by ghost** [comment]
HI @sshymko thank you for you report, i' closing this on comment -> https://github.com/magento/magento2/issues/21106#issuecomment-462460396

**Comment by sshymko** [comment]
@engcom-backlog-nazar 
Why was the ticket closed?
The issue with the `loops=1` is reproducible on vanilla Magento of any version.
Yes, there's a workaround to use `loops=2`, but the bug report is still valid.
Please reopen the ticket.

**Comment by ghost** [comment]
@sshymko so how we can fix this in magento ? if problem not in  magento codebase ?

**Comment by pmouawad** [comment]
@engcom-backlog-nazar, maybe you should open another issue to fix issues in test plan.

See https://github.com/magento/magento2/issues/21106#issuecomment-462465899 and other performance issues.

Thanks

**Comment by ghost** [comment]
Hi @pmouawad thanks as this file located in setup module we can leave this issue here, thanks for explanation

**Comment by sshymko** [comment]
@engcom-backlog-nazar 
JMeter JMX file is in the Magento codebase and it's implemented in such a way that it's impossible to run scenarios once (loops = 1). The fix should be in the JMX logic. It could be as dirty as `loops = loops + 1`. From the user standpoint, all input parameters should behave as intended.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-backlog-nazar 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-98332`, `MAGETWO-98333` were created

**Issue Available**: @engcom-backlog-nazar, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
Hi @sshymko. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-24354` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.
